### PR TITLE
Make 2D and 3D node names more explicit

### LINKED ISSUE #30736: Proposal for renaming nodes for Godot 4.0
## Breaking compatibility

I know I promised to break as minimum as possible to compatibility for Godot 4.0, but I' ve been thinking this over for a while and based on feedback, and I think it's worth it. It should not break compatibility with existing scenes (we can add internal renames), but it will with code...

So maybe what we could do is a tool in Godot 4.0 that basically converts a script from 3.0 to 4.0 doing symbol renaming or just re-tokening.. this way no manual work has to be done by the user (so no huge amounts of pain like when we moved from 2.0 to 3.0).

For reusing existing tutorials (that I think won't really be affeced much anyway), we can put up an doc article showing what was renamed.

## What I would change

Despite that many say that Godot was originally a 2D engine, that is false. Originally, it was a 3D engine and the only things supported as 2d where the canvas (UI) nodes. The 2D nodes came later.

So, this makes it quite confusing that for 3D nodes, the "Spatial" term is used and for 2D nodes, there is "Node2D".

For users starting to use Godot (and even for advanced users), this is pretty confusing in general (only aided by the node colors, but if you make a mistake in code, it can be confusing to realize you are using the right version 3D instead of 2D)).

My proposal is to simply make explicit, in case nodes exist both as 2D and 3D form, that they should be named the same, but add a 3D or 2D at the end (save for some exceptions).

So, typical renames would be:

* Spatial -> Node3D
* Area -> Area3D
* Camera -> Camera3D
* Navigation -> Nagivagion3D
* RemoteTransform -> RemoteTransform3D
* KinematicBody -> KinematicBody3D

and so on.

For some, I think we could keep the current way, because the use case is mainly for 2D or 3D, and the opposite version, such as:

* Sprite and Sprite3D make sense as is, since Sprite is mainly a 2D node
* MeshInstance / MeshInstance2D makes sense, because meshes are mainly used for 3D

But then again, feel free to suggest if you would prefer something more explicit and make everything 2D/3D always.

## Namespaces

I'm not that too keen on namespaces within Godot for a few reasons
* Godot is an application, not a library
* Everything has a a clear role when you look at inheritance.
* Debug symbols increase considerably with namespaces, and our debug builds will be much larger

But nothing should avoid us from defining them using the ClassDB API, so languages that are heavily namespace based, and where users are very used to it (like C#) can make use of them.

As an example: 

Below a GDCLASS("Node2D") definition, an optional
GDNAMESPACE("Nodes2D")

It will also make browsing help and documentation probably easier too.

## Other things I would like to change

SpatialMaterial is confusing as hell to new users (and existing too). I think we should rename to StandardMaterial3D. I would probably also want to make two different versions, one similar to the existing one, and another that instead uses ORM style textures (setting them up in Godot right now is a huge hassle using Spatial Material and assigning channels manually). This way we could have something like:

```
+ Material3D
- + StandardMaterial3D
- + ORMMaterial3D
```
or similar..

The 3D scene import process is still quite a hassle because you can't set options to individual meshes and materials while importing them, so I want to add an option in the import dock to have import windows with settings for some types of resources (in this case for imported scenes). 

The 3D scene import would show you a tree with all the data and then you can choose a lot of nice stuff in there for every node/material such as:

* Keep material built-in
* Replace this material by this file
* Edit/Create Mesh LODs for proper level of detail
* Edit mesh lightmapping options, including lightmap scale so you can have different scales in the lightmap for different objects, etc
* Many other options.

Feedback welcome!

### PR DESCRIPTION
**Pull request by reduz**
To avoid confusion from users, most 3D nodes and 2D nodes (as well as resources) were renamed to match their type, as an example:

* Area -> Area3D
* MeshInstance -> MeshInstance3D
* Sprite -> Sprite2D

Additionally, "Spatial" as a base for 3D nodes no longer exists, it was renamed to "Node3D".

Some nodes that are obviously 3D or 2D only remain as such, such as GIProbe, or YSort, but far most of them were renamed.

Compatibility option was added so old scenes can still load without problem.

Below is the list of 3D nodes and their new names:
image

*Bugsquad edit:* Fixes #30736

### CODE REVIEW

**Comment by aaronfranke** [comment]
Shouldn't `VehicleBody` also have a `3D` suffix? While Godot may only have 3D vehicle physics right now, it's completely feasible for there to be 2D vehicles.

**Comment by bojidar-bg** [comment]
Likewise for ClippedCamera, shouldn't it be ClippedCamera3D? In the end, it is a clipped "Camera3D" now, so I don't see why the suffix should be dropped.

**Comment by akien-mga** [comment]
Nodes left to consider (probably left alone on purpose, but we should confirm that we don't want to rename them):

- In `scene/2d` and inheriting Node2D:
  * BackBufferCopy
  * CanvasModulate
  * ParallaxLayer
    - ParallaxBackground inherits Node but its logic depends on Node2D, Camera2D
  * TileMap
  * TouchScreenButton

- In `scene/3d` and inheriting Node3D:
  * All ARVR nodes - but that was left for @BastiaanOlij probably, and they're meant to also be renamed to change ARVR to XR. Should they get 3D suffix too? (e.g. XRCamera3D)
  * BakedLightMap (dead code, kept for future reimplementation)
  * GIProbe
  * ReflectionProbe
  * VehicleBody, VehicleWheel
  * WorldEnvironment: inherits Node but its logic depends on Node3D

Nodes that could maybe be moved:

- CanvasItem is in `scene/2d`, but only depends on `scene/main` stuff. Should maybe be moved in `main`? Especially since it's also used by `scene/gui`.
- SkeletonIK3D is in `scene/animation`, but it depends on 3D nodes. Should it move to `scene/3d`?

**Comment by akien-mga** [comment]
> Likewise for ClippedCamera, shouldn't it be ClippedCamera3D? In the end, it is a clipped "Camera3D" now, so I don't see why the suffix should be dropped.

Already renamed locally.

**Comment by akien-mga** [comment]
Added more renames, namely:
- ClippedCamera -> ClippedCamera3D
- HeightMapShape -> HeightMapShape3D
- Particles2D -> GPUParticles2D
- World -> World3D

Also renamed editor plugins to match the new node names that they handle, and fixed a few missing file renamings to match their new main class name.

**Comment by aaronfranke** [comment]
> TouchScreenButton

Why isn't this `Control` derived? Disclaimer: I haven't looked at what this actually is, but I would expect buttons to all be GUI elements derived from `Control`.

> All ARVR nodes

There is no possibile way to have non-3D ARVR nodes, so I don't think they should get the suffix.

> WorldEnvironment: inherits Node but its logic depends on Node3D

If that is the case, why does its icon contain both red and blue? Blue is for 2D.

> CanvasItem is in `scene/2d` ... also used by `scene/gui`

Not a suggestion, just a thought: What about `scene/canvas/2d` and `scene/canvas/gui` (or `scene/canvas/control` to make the folder structure match the node inheritance?

Also, there is the case of non-node things to rename (such as `Transform` -> `Transform3D`).

File: editor/node_3d_editor_gizmos.h [hunk 1]
```diff
@@ -31,7 +31,7 @@
 #ifndef SPATIAL_EDITOR_GIZMOS_H
 #define SPATIAL_EDITOR_GIZMOS_H
```
Comment (aaronfranke) [inline editor/node_3d_editor_gizmos.h]: Should also change the include guards along with the class and file name.

Comment (akien-mga) [reply editor/node_3d_editor_gizmos.h]: Too difficult to do consistently with all those renames, I'll write a script to do it (#37290).

**Comment by reduz** [comment]
@aaronfranke touchscreenbutton is not control derived simply because it will process its own input (touch). its completely unrelated to any of the control logic (mouse).

**Comment by akien-mga** [comment]
> Nodes that could maybe be moved:
> 
>     * CanvasItem is in `scene/2d`, but only depends on `scene/main` stuff. Should maybe be moved in `main`? Especially since it's also used by `scene/gui`.
> 
>     * SkeletonIK3D is in `scene/animation`, but it depends on 3D nodes. Should it move to `scene/3d`?

Done.

> In `scene/3d` and inheriting Node3D:
>     * VehicleBody, VehicleWheel

Done.

**Comment by Calinou** [comment]
@aaronfranke In Godot 3.2 at least, WorldEnvironment can be used for glow and HDR in 2D. I'm not sure about the `master` branch though.

**Review by akien-mga — APPROVED** [review]
Should be good to merge as far as I'm concerned if the CI builds pass.

**Comment by akien-mga** [comment]
Pushed some extra changes for leftover String references to old node names, and fixing the editor plugin names for some of the renamed node types.

**Comment by Demolishun** [comment]
Will there be a way to automatically update a project to the new names?

**Comment by Anutrix** [comment]
Just a note: Docs and header gaurds need to be updated.

**Comment by danilw** [comment]
Vector4 now cry alone 😢

**Comment by aaronfranke** [comment]
@danilw That's not related to this PR. If you want to discuss that, check out this discussion: https://github.com/godotengine/godot-proposals/issues/629

**Comment by Calinou** [comment]
@Demolishun This is already implemented:

> Compatibility option was added so old scenes can still load without problem.

It works automatically, you don't need to do anything specific.

**Pull request merged**